### PR TITLE
Execute strategySize first in write()

### LINKED ISSUE #675: strategy.size() can reenter WritableStream logic before its caller finishes all the work
Looks ReadableStream has the same issue.

The other user provided functions e.g. ones on underlyingSource/Sink which are invoked after their caller finishes all the work that touches the state variables of the streams. But in the callers of strategy.size(), there're lines that touch the state variables (both reading and writing) after strategy.size() is invoked. strategy.size() can implemented to call writer function or controller functions, and therefore reentry may happen. Should we have some guard against such a reentry?

### PR DESCRIPTION
**Pull request by ricea**
A strategy size() function can re-entrantly call controller, writer or writable
methods, leaving the stream in an unexpected state. To ensure predictable
behaviour, call size() before observing or modifying any state.

Previously, if size() threw an exception, writer.write() would reject with that
exception. After this change, it rejects with a TypeError instead.

Fixes #675.

### CODE REVIEW

File: reference-implementation/lib/writable-stream.js [hunk 1]
```diff
@@ -654,7 +644,17 @@ function WritableStreamDefaultWriterRelease(writer) {
...
 
-  assert(stream !== undefined);
+  if (stream === undefined) {
+    return Promise.reject(defaultWriterLockException('write to'));
+  }
+
+  const controller = stream._writableStreamController;
+
+  const chunkSize = WritableStreamDefaultControllerGetChunkSize(controller, chunk);
+
```
Comment (ricea) [inline reference-implementation/lib/writable-stream.js]: Drive-by comment: maybe we should check that `stream === writer._ownerWritableStream` before proceeding?

strategySize might have
1. Detached the writer
2. Created a new writer and done a bunch of writes with it

It's probably better just to eliminate the possibility rather than try to work out whether it would be a problem or not.

Needs test(s).

File: reference-implementation/lib/writable-stream.js [hunk 2]
```diff
@@ -505,16 +505,6 @@ class WritableStreamDefaultWriter {
       return Promise.reject(defaultWriterBrandCheckException('write'));
     }
 
-    const stream = this._ownerWritableStream;
```
Comment (domenic) [inline reference-implementation/lib/writable-stream.js]: I am sad about moving these checks as I like the separation of invariant-checking in the public API only. It seems like at least the first check, of stream being undefined, can stay here. What about the second one?

File: reference-implementation/lib/writable-stream.js [hunk 3]
```diff
@@ -747,6 +747,21 @@ function WritableStreamDefaultControllerClose(controller) {
   WritableStreamDefaultControllerAdvanceQueueIfNeeded(controller);
 }
 
+function WritableStreamDefaultControllerGetChunkSize(controller, chunk) {
```
Comment (domenic) [inline reference-implementation/lib/writable-stream.js]: This appears to be called from only one place, so inlining it fits better with our style.

File: reference-implementation/lib/writable-stream.js (hunk 1 above)
Comment (ricea) [inline reference-implementation/lib/writable-stream.js]: I added tests to my wpt change and added the extra check here.

File: reference-implementation/lib/writable-stream.js (hunk 3 above)
Comment (ricea) [inline reference-implementation/lib/writable-stream.js]: Originally I didn't inline it because it was a controller operation rather than a writer operation.

Afterwards I realised that once we have a ByteController this will become a polymorphic method call. Maybe the "you ain't gonna need it" principle applies, but I'm pretty sure that we will add a byte controller and that this separation will make sense at that time.

File: reference-implementation/lib/writable-stream.js (hunk 2 above)
Comment (ricea) [inline reference-implementation/lib/writable-stream.js]: I moved the stream check back into write().

We need to check `state` after strategy.size() returns, since it may have changed. We could additionally check it before, if we wanted to guarantee that size() is only called in the writable state, but there are no tests for that and I didn't consider it important.

I have started thinking of size() as part of the public API. This makes the line between public methods and abstract operations a bit fuzzier.

File: reference-implementation/lib/writable-stream.js [hunk 4]
```diff
@@ -656,6 +650,18 @@ function WritableStreamDefaultWriterWrite(writer, chunk) {
...
 
+  const controller = stream._writableStreamController;
+
+  const chunkSize = WritableStreamDefaultControllerGetChunkSize(controller, chunk);
+
+  if (stream !== writer._ownerWritableStream) {
+    return Promise.reject(defaultWriterLockException('write to'));
+  }
+
+  if (stream._state === 'closing') {
```
Comment (domenic) [inline reference-implementation/lib/writable-stream.js]: I guess this could be consolidated with line 666.

**Comment by domenic** [comment]
Spec changes LGTM too. Will let you coordinate landing WPTs and then this.

File: reference-implementation/lib/writable-stream.js [hunk 5]
```diff
@@ -656,20 +650,26 @@ function WritableStreamDefaultWriterWrite(writer, chunk) {
 
   assert(stream !== undefined);
 
+  const controller = stream._writableStreamController;
+
+  const chunkSize = WritableStreamDefaultControllerGetChunkSize(controller, chunk);
```
Comment (tyoshino) [inline reference-implementation/lib/writable-stream.js]: It's really unfortunate that we need to do this in the Writer's abstract op. Ideally, the detail of the operation should be hidden. But ok for now.
